### PR TITLE
Task report finish to job manager

### PR DESCRIPTION
**Pull request by liuyu85cn**
What problem this pr want to address: 

when JobManager(meta) dispatch tasks to storage, 
it will call "collectAll" to wait all task finished.
if meta leader change during this collectAll, 
meta will fail to write job status to kvstore, forever.

So, don't let meta wait for those tasks. Instead of when a single task finished, 
task manager(storage) will call reportTaskFinish to the leader of meta.

### CODE REVIEW

File: src/meta/processors/jobMan/JobManager.cpp [hunk 1]
```diff
@@ -79,10 +79,10 @@ void JobManager::shutDown() {
 
 void JobManager::scheduleThread() {
     LOG(INFO) << "JobManager::runJobBackground() enter";
-    while (status_ == Status::RUNNING) {
+    while (status_ != JbmgrStatus::STOPPED) {
```
Comment (panda-sheep) [inline src/meta/processors/jobMan/JobManager.cpp:82]: why not while (status_ == JbmgrStatus::IDLE || status_ == JbmgrStatus::BUSY) ？

File: src/meta/processors/jobMan/JobManager.cpp [hunk 2]
```diff
@@ -79,10 +79,10 @@ void JobManager::shutDown() {
...
         int32_t iJob = 0;
-        while (!try_dequeue(iJob)) {
-            if (status_ == Status::STOPPED) {
+        while (status_ == JbmgrStatus::BUSY || !try_dequeue(iJob)) {
```
Comment (panda-sheep) [inline src/meta/processors/jobMan/JobManager.cpp:84]: the status_ here can be IDLE?

**Review by panda-sheep — COMMENTED** [review]
Good job!

File: src/meta/processors/jobMan/JobManager.cpp (hunk 2 above)
Comment (panda-sheep) [reply src/meta/processors/jobMan/JobManager.cpp:84]: my fault you are right

File: src/meta/processors/jobMan/StatisJobExecutor.cpp [hunk 3]
```diff
@@ -57,6 +57,56 @@ StatisJobExecutor::executeInternal(HostAddr&& address, std::vector<PartitionID>&
                                  std::move(parts), concurrency_, &(statisItem_[address]));
 }
 
+void showStatisItem(const cpp2::StatisItem& item, const std::string& msg) {
```
Comment (panda-sheep) [inline src/meta/processors/jobMan/StatisJobExecutor.cpp:60]: What is this interface for?

File: src/meta/processors/jobMan/StatisJobExecutor.cpp [hunk 4]
```diff
@@ -57,6 +57,56 @@ StatisJobExecutor::executeInternal(HostAddr&& address, std::vector<PartitionID>&
...
+ */
+cpp2::ErrorCode StatisJobExecutor::saveSpecialTaskStatus(const cpp2::ReportTaskReq& req) {
+    if (!req.__isset.statis) {
+        return cpp2::ErrorCode::SUCCEEDED;
+    }
+    cpp2::StatisItem statisItem;
+    auto statisKey = MetaServiceUtils::statisKey(space_);
+    std::string val;
+    auto ret = kvstore_->get(kDefaultSpaceId, kDefaultPartId, statisKey, &val);
+    if (ret == kvstore::ResultCode::SUCCEEDED) {
```
Comment (panda-sheep) [inline src/meta/processors/jobMan/StatisJobExecutor.cpp:101]: There maybe be a problem here:
If the last statis job and the current statis job are in the same space,
do you add the results of the last time?

File: src/meta/processors/jobMan/StatisJobExecutor.cpp (hunk 3 above)
Comment (liuyu85cn) [reply src/meta/processors/jobMan/StatisJobExecutor.cpp:60]: for debug ... and I think it maybe used one day. If you don't like it, I'm ok to remove those.

File: src/meta/processors/jobMan/StatisJobExecutor.cpp (hunk 4 above)
Comment (liuyu85cn) [reply src/meta/processors/jobMan/StatisJobExecutor.cpp:101]: discussed offline, The previous status value will be init at prepare state.

File: src/meta/processors/jobMan/JobManager.cpp [hunk 5]
```diff
@@ -91,78 +85,210 @@ void JobManager::scheduleThread() {
...
+    auto jobExec =
+        MetaJobExecutorFactory::createMetaJobExecutor(*optJobDesc, kvStore_, adminClient_);
+
+    if (jobExec) {
+        if (!optJobDesc->getParas().empty()) {
+            auto spaceName = optJobDesc->getParas().front();
+            auto spaceId = getSpaceId(spaceName);
+            LOG(INFO) << folly::sformat("spaceName={}, spaceId={}", spaceName, spaceId);
+            if (spaceId != -1) {
+                jobExec->setSpaceId(spaceId);
```
Comment (critical27) [inline src/meta/processors/jobMan/JobManager.cpp:174]: Maybe we could add a spaceId in thrift `AdminJobReq`, then the paras could just save their own parameters later. And job description could directly use it when job/task added.

File: src/meta/processors/jobMan/StatisJobExecutor.cpp [hunk 6]
```diff
@@ -57,51 +64,83 @@ StatisJobExecutor::executeInternal(HostAddr&& address, std::vector<PartitionID>&
...
+
+    lhs.space_vertices += rhs.space_vertices;
+    lhs.space_edges += rhs.space_edges;
+    // ignore part_corelativity, they shoule be same
+}
+
+/**
+ * @brief caller will guarantee there won't be any conflict read / write.
+ */
+cpp2::ErrorCode StatisJobExecutor::saveSpecialTaskStatus(const cpp2::ReportTaskReq& req) {
```
Comment (critical27) [inline src/meta/processors/jobMan/StatisJobExecutor.cpp:100]: Do we need to add lock for accumulate result?

File: src/meta/processors/jobMan/JobManager.h [hunk 7]
```diff
@@ -113,13 +134,13 @@ class JobManager : public nebula::cpp::NonCopyable, public nebula::cpp::NonMovab
     // The job in running or queue
     folly::ConcurrentHashMap<JobID, JobDescription>    inFlightJobs_;
 
-    std::unique_ptr<thread::GenericWorker>             bgThread_;
+    std::thread                                        bgThread_;
```
Comment (critical27) [inline src/meta/processors/jobMan/JobManager.h:137]: What is the issue of `GenericWorker`?

File: src/meta/processors/jobMan/StatisJobExecutor.cpp (hunk 6 above)
Comment (liuyu85cn) [reply src/meta/processors/jobMan/StatisJobExecutor.cpp:100]: no, like I said in comments.

Comment (critical27) [reply src/meta/processors/jobMan/StatisJobExecutor.cpp:100]: 👌

File: src/meta/processors/jobMan/JobManager.h (hunk 7 above)
Comment (liuyu85cn) [reply src/meta/processors/jobMan/JobManager.h:137]: got some weird dead lock. And I believe I just need a simple thread to run "scheduleThread".

Comment (critical27) [reply src/meta/processors/jobMan/JobManager.h:137]: 👌

File: src/meta/processors/jobMan/JobManager.cpp (hunk 5 above)
Comment (liuyu85cn) [reply src/meta/processors/jobMan/JobManager.cpp:174]: ok, and I think in another pr is better, in which I can change all the existing job

File: src/meta/processors/jobMan/JobManager.cpp (hunk 1 above)
Comment (liuyu85cn) [reply src/meta/processors/jobMan/JobManager.cpp:76]: because it's long ...

**Review by critical27 — APPROVED** [review]
LGTM

**Review by bright-starry-sky — APPROVED** [review]
LGTM

**Pull request merged**